python/spark_sklearn/tests/test_gapply.py

```diff
@@ -0,0 +1,216 @@
+        ordered = converted.sort_values(df.columns.tolist())
+        # We need to drop the index after sorting because pandas remembers the pre-sort
+        # permutation in the old index. This would trigger a failure if we were to compare
+        # differently-ordered dataframes, even if they had the same sorted content.
+        unindexed = ordered.reset_index(drop=True)
+        return unindexed
+    actual = normalize(actual)
+    expected = normalize(expected)
+    pd.util.testing.assert_frame_equal(actual, expected)
+
+def _emptyFunc(key, vals):
+    return pd.DataFrame.from_records([])
+
+@fixtureReuseSparkSession
+class GapplyTests(unittest.TestCase):
+    
+    def test_gapply_empty(self):
+        # Implicitly checks that pandas version is large enough (unit tests for the actual version
+        # checking itself would require some serious mocking)
+        longLongSchema = StructType().add("a", LongType()).add("b", LongType())
+        emptyLongLongDF = self.spark.createDataFrame([], schema=longLongSchema)
+        gd = emptyLongLongDF.groupBy("a")
+        self.assertEqual(gapply(gd, _emptyFunc, longLongSchema, "b").collect(), [])
+
+    @unittest.skip("Generating an empty dataframe after exploding fails, see SPARK-16179")
+    def test_gapply_empty_schema(self):
+        longLongSchema = StructType().add("a", LongType()).add("b", LongType())
+        emptyLongLongDF = self.spark.createDataFrame([(1, 2)], schema=longLongSchema)
+        gd = emptyLongLongDF.groupBy("a")
+        self.assertEqual(gapply(gd, _emptyFunc, StructType(), "b").collect(), [])
+
+    def test_gapply_raises_if_bad_schema(self):
+        longLongSchema = StructType().add("a", LongType()).add("b", LongType())
+        emptyLongLongDF = self.spark.createDataFrame([], schema=longLongSchema)
+        gd = emptyLongLongDF.groupBy("a")
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, LongType(), "b")
+        
+    def test_gapply_raises_if_bad_cols(self):
+        longLongLongSchema = StructType() \
+                             .add("a", LongType()).add("b", LongType()).add("c", LongType())
+        emptyLongLongLongDF = self.spark.createDataFrame([], schema=longLongLongSchema)
+        gd = emptyLongLongLongDF.groupBy("a")
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, longLongLongSchema, "b", "*")
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, longLongLongSchema, "*")
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, longLongLongSchema, "b", "b")
+        self.assertRaises(ValueError, gapply, gd, _emptyFunc, longLongLongSchema, "b", "c", "b")
+
+    NROWS = 15
+    NKEYS = 5
+    NVALS = 100
+
+    def checkGapplyEquivalentToPandas(self, pandasAggFunction, dataType, dataGen):
+        schema = StructType().add("val", dataType)
+        pandasDF = pd.DataFrame.from_dict({
+            "key": [random.randrange(GapplyTests.NKEYS) for _ in range(GapplyTests.NROWS)],
```

Comment:
No need to do that. We tend to fix a seed in the beforeAll and we do not bother with time-based.